transaction/beef.go

```diff
@@ -920,17 +1018,66 @@ func (b *Beef) Bytes() ([]byte, error) {
 
 	// transactions / txids
 	beef = append(beef, VarInt(len(b.Transactions)).Bytes()...)
-	for _, tx := range b.Transactions {
-		beef = append(beef, byte(tx.DataFormat))
+	txs := make(map[string]struct{}, len(b.Transactions))
+	var appendTx func(tx *BeefTx) error
+	appendTx = func(tx *BeefTx) error {
+		if _, ok := txs[tx.Transaction.TxID().String()]; ok {
+			return nil
+		}
+
 		if tx.DataFormat == TxIDOnly {
+			beef = append(beef, byte(tx.DataFormat))
 			beef = append(beef, tx.KnownTxID[:]...)
 		} else {
+			for _, txin := range tx.Transaction.Inputs {
+				if parentTx := b.findTxid(txin.SourceTXID.String()); parentTx != nil {
+					if err := appendTx(parentTx); err != nil {
+						return err
+					}
+				}
+			}
+			beef = append(beef, byte(tx.DataFormat))
 			if tx.DataFormat == RawTxAndBumpIndex {
 				beef = append(beef, VarInt(tx.BumpIndex).Bytes()...)
 			}
 			beef = append(beef, tx.Transaction.Bytes()...)
 		}
+		txs[tx.Transaction.TxID().String()] = struct{}{}
+		return nil
+	}
+	for _, tx := range b.Transactions {
+		if err := appendTx(tx); err != nil {
+			return nil, err
+		}
 	}
 
 	return beef, nil
 }
+
+func (b *Beef) AtomicBytes(txid *chainhash.Hash) ([]byte, error) {
+	beef, err := b.Bytes()
+	if err != nil {
+		return nil, err
+	}
+	return append(append(util.LittleEndianBytes(ATOMIC_BEEF, 4), txid[:]...), beef...), nil
```

Comment:
Since the final size is known, consider reserving memory before append.

```suggestion
	result := make([]byte, 0, 4+chainhash.HashSize+len(beef))
	result = append(result, util.LittleEndianBytes(ATOMIC_BEEF, 4)...)
	result = append(result, txid[:]...)
	result = append(result, beef...)

	return result, nil
```